### DISCUSSION TITLE
Removing the super-table from `toToml` output

### DISCUSSION THREAD

**Discussion by bradenhilton** [twpayne/chezmoi#2539] [Show and tell]
If you use the [`toToml` template function](https://www.chezmoi.io/reference/templates/functions/toToml/) with nested `dict`s, you may notice the super-table in the output:

```console
❯ chezmoi execute-template '{{ dict "key1" (dict "key2" (dict "values" (list "value1" "value2"))) | toToml }}'
[key1]
[key1.key2]
values = ['value1', 'value2']
```

The super-table[ is unnecessary](https://toml.io/en/v1.0.0#table):

> You don't need to specify all the super-tables if you don't want to. TOML knows how to do it for you.
> 
> ```toml
> # [x] you
> # [x.y] don't
> # [x.y.z] need these
> [x.y.z.w] # for this to work
> 
> [x] # defining a super-table afterward is ok
> ```

To remove the super-table, you can combine `splitList`, `rest` and `join` like so to remove the super-table line (requires one template action `{{ ... }}` per super-table `dict`):

```console
❯ chezmoi execute-template '{{ dict "key1" (dict "key2" (dict "values" (list "value1" "value2"))) | toToml | splitList "\n" | rest | join "\n" }}'
[key1.key2]
values = ['value1', 'value2']
```

**Comment by twpayne** [comment]
> To remove the super-table, you can combine `splitList`, `rest` and `join` like so:

This removes the first line of the output, not all super-tables.

**Reply by bradenhilton** [reply]
Correct. I was working under the assumption that one template action per super-table would be used.

I don't really know of a better way to handle this given that we depend on `pelletier/go-toml`.

**Reply by bradenhilton** [reply]
I believe I have committed a great RegEx atrocity.

```
{{- $input := dict 
    "key1" (dict
        "key2" (dict "values" (list "value1" "value2"))
        "key3" (dict "values" (list "value1" "value2"))
        )
    | toToml 
-}}
{{- $nestedKeys := regexFindAll "\\[(.+)\\..*\\]" $input -1 -}}

{{- $superKeys := list -}}

{{- range $_, $key := $nestedKeys -}}
{{-  $superKeys = append $superKeys (regexFind "[^\\[].+?[^\\.][^\\]]" $key) -}}
{{- end -}}

{{- range $_, $superKey := ($superKeys | uniq) -}}
{{-  $input = $input | replaceAllRegex (printf "\\[%s\\]\\n" $superKey) "" -}}
{{- end -}}

{{- $input -}}
```

```
[key1.key2]
values = ['value1', 'value2']

[key1.key3]
values = ['value1', 'value2']
```

**Reply by halostatue** [reply]
I believe that this has many more than [two problems](http://regex.info/blog/2006-09-15/247).

**Reply by halostatue** [reply]
This feels like something that may not be necessary, but could potentially be an option to `toToml` (maybe `{{ dict | toToml "compact" }}`) and prevent others from reimplementing this. This would be matchable with `(?m)^(^\[[^\]]+\])$\[[^\]]+\]$` in Go, I think.

**Reply by bradenhilton** [reply]
It's easier in code for sure. For one, I don't think the template functions support capture groups, hence all the intermediate steps. The example I posted doesn't handle `.`s within quotes (`'a.b'.c`) as well.

The super-tables also don't alter the functionality of the file, but having the option to remove them can help with `modify_` templates. Similarly, also having the ability to select between single and double quotes would be ideal, but I don't think `pelletier/go-toml` supports either of them.

**Reply by halostatue** [reply]
`pelletier/go-toml` explicitly only supports single quotes unless the string requires double quotes (v2 is *much* more opinionated than the first version). It looks like @twpayne switched to directly using `pelletier/go-toml` [three years ago](https://github.com/twpayne/chezmoi/blame/987450bff1853c97312c25ca9b3c8f33f2f764dc/go.mod) and removed the previous dependency `BurntSushi/toml`.

I just [tested BurntSushi/toml](https://go.dev/play/p/aToyCGLqBjR) and it has the same output as `pelletier/go-toml`, except indented (and it does not appear to have the same configuration flexibility as `pelletier/go-toml`).

Unless we're going to post-modify TOML output (which I don’t really think is a good idea, despite my suggestions above). The parent keys aren't necessary, but *are* legitimately in spec.

**Reply by twpayne** [reply]
> Unless we're going to post-modify TOML output (which I don’t really think is a good idea, despite my suggestions above). The parent keys aren't necessary, but _are_ legitimately in spec.

+1 on this.

Note also that the generated config file is for machine consumption, so we shouldn't care too much about removing in-spec behavior. The `fromToml` / `toToml` round trip also removes comments so the resulting file isn't particularly human-friendly.

**Reply by bradenhilton** [reply]
I'm unsure if there is a prevailing impression that I was making a feature request but this is not the case.

This is an extremely niche use-case, and I deliberately posted this in "Show and tell" in the event that anyone else was looking to accomplish the same and didn't know where to start.

**Reply by halostatue** [reply]
No, I knew that this is *musing* far more than anything else. But the musing is still interesting.
